Media stream token management

jmartin-sul opened the pull request:
PR #44 should probably be merged first, then this should get rebased on top of that.

* add a StacksMediaToken class for ephemeral authorization of access by a user to a StacksMediaStream.
* have MediaController#stream include an encrypted token string in its redirects.
* add MediaController#verify_token method (and route to it) to allow other services to verify tokens.

resolve #28

ndushay commented on app/controllers/media_controller.rb:
```diff
@@ -19,14 +19,24 @@ def stream
 
+  # rubocop:disable Metrics/LineLength
+  def verify_token
+    if StacksMediaToken.verify_encrypted_token_string params[:token_string], params[:id], params[:file_name], params[:user_ip_addr]
```
if params[:blah]  is used in multiple places in the file, it may make sense to pull out a private method

```
def file_name
  @fname ||= (allowed_)params[:file_name]
end
```
which could also alleviate the line that is too long.

Another way to alleviate that long line would be to make the conditional statement a private method?

jmartin-sul replied:
sure, that seems reasonable.

i figured the `allowed_` part would get added when i rebased on master after the strong params PR got merged.

ndushay replied:
yup - allowed_params will become avail w rebasing (and should always be used).  That's why I put it in parens.  :-)

ndushay commented on app/controllers/media_controller.rb:
```diff
@@ -46,4 +56,12 @@ def media_params
     @media ||= StacksMediaStream.new(media_params)
   end
+
+  def remote_ip_addr
```
is this the user ip, the wowza ip, or "it depends"?

jmartin-sul replied:
this is the user IP, because `remote_ip_addr` is called when the token is being created, at which point the user is talking directly to stacks.

in `verify_token`, it's also the user IP, but it's given to stacks in a URL param, since wowza's making the request to stacks, at that point.

i can add a comment, that could use some clarification in the code.

ndushay commented on app/models/stacks_media_token.rb:
```diff
@@ -0,0 +1,91 @@
+# * this checking in wowza can be implemented via a wowza request processing plugin.
+#
+# note that there's no check for whether a token's been used (there's no record of the token
+# minting in memory or DB, and so nothing to mark).
```
don't we want a token expiry time as well?  As there is for the IIIF token?

jkeck replied:
@ndushay is [this](https://github.com/sul-dlss/digital_stacks_rails/pull/45/files#diff-1593072322c6911944310a140001ba4eR86) what you're looking for?

ndushay replied:
Maybe I'm not understanding the code, but with MAX_TOKEN_AGE set to 1 second, where are we evaluating if the token is no more than x minutes old?   Or is 1 second meant to be enough time?

As I understood the tests, they were only checking that the timestamp of the token is between the beginning of running the tests and the evaluation of the token.

ndushay commented on spec/controllers/media_controller_spec.rb:
```diff
@@ -79,4 +85,43 @@
+    let(:user_ip_addr) { '192.168.1.100' }
+    let(:token) { StacksMediaToken.new(id, file_name, user_ip_addr) }
+
+    # rubocop:disable Metrics/LineLength
```
personally, I hate inline rubocop exceptions.  Can the test be written in a way to avoid the long lines?   if we MUST have them, can the inline exceptions be limited to only those tests that really need them?   Or can we use a file exception in rucobop.yml?    @jkeck probably has a differing opinion.

jkeck replied:
The departmental best practice is to use the `.rubocop_todo.yml` file (possibly using rubocop's `--auto-gen-config`).

I would be interested if there would be a way to write these tests (and related code) in a way that fell within the line-length limitation (with 120 being pretty generous).

@jmartin-sul, more than happy to chat or pair on ways to do that while keeping the code readable if you would like.

jmartin-sul replied:
@jkeck:  yup, happy to pair/chat to come up with a way to shorten lines.  i couldn't think of anything offhand that didn't detract from readability or add what seemed like unnecessary temp vars.  i'm happy to go with whatever people want to do on line length.

@ndushay:  agreed, i also really dislike inline rubocop exceptions, but i saw an empty `.rubocop_todo.yml` and other parts of the code that had inline exceptions, so i figured i'd just stay consistent with the codebase.

anyway, i often find longer lines to be more readable than the things people do to avoid them, but again, not a cause i feel the need to fight for.

ndushay commented on spec/models/stacks_media_token_spec.rb:
```diff
@@ -0,0 +1,59 @@
+    end
+  end
+
+  # rubocop:disable Metrics/LineLength
```
ditto comments about inline rubocop exceptions.

ndushay commented:
My main concern is that we don't seem to be doing anything re: token expiration.

ndushay commented on app/models/stacks_media_token.rb:
```diff
@@ -0,0 +1,91 @@
+      unverified_smt.id == id &&
+      unverified_smt.file_name == file_name &&
+      unverified_smt.user_ip_addr == user_ip_addr &&
+      (unverified_smt.timestamp >= MAX_TOKEN_AGE.ago)
```
to my eyes, this is the only check for the timestamp validity -- is it newer than 1 second ago.  Is that a reasonable time frame?

jmartin-sul replied:
yeah, this is rails magic that i imagine is intended to be natural language-y, but it ends up being a bit opaque.  but `MAX_TOKEN_AGE.ago` actually returns a time stamp, because if you call the `.ago` method on a duration, it'll give you a date.  e.g.:

```
DN0a23226d:digital_stacks_rails suntzu$ rails c
Loading development environment (Rails 4.2.5.2)
2.2.2 :001 > 1.second
 => 1 second 
2.2.2 :002 > 1.second.ago
 => Mon, 02 May 2016 22:04:42 UTC +00:00 
```

so this actually does check expiry.

i believe i ripped the idea off from here:  https://github.com/sul-dlss/digital_stacks_rails/blob/046a08d944487ff2612b9cea6c723747ca5347fc/app/models/user.rb#L31

FWIW, i saw your comment and was worried that i had made the oversight you were concerned about, and had to double-check that all over again.  so that seems like a good indication that a comment might be helpful.

ndushay commented on app/models/stacks_media_token.rb:
```diff
@@ -0,0 +1,91 @@
+# note that there's no check for whether a token's been used (there's no record of the token
+# minting in memory or DB, and so nothing to mark).
+class StacksMediaToken
+  MAX_TOKEN_AGE = 1.second
```
Perhaps the MAX_TOKEN_AGE value should be in the Settings yml?  1 second for tests, but xxx recommended for in situ?

jmartin-sul replied:
good call, will move to settings

jmartin-sul commented:
ok, i believe the above concerns have been addressed:
* inline rubocop exceptions have been removed, because lines are now all <= 120 chars in length.
* comments added on the confusing `remote_ip_addr` vs `params[:user_ip_addr]` thing in `MediaController`.
* comments added about the expiry check.  when i was there to add the comment, i noticed a bit of refactoring i could do to make things a bit more modular and testable.
* and as a result of that refactor, it was easy to add a test for naomi's legit concern about lack of testing for expired tokens (which is now specifically tested [here](https://github.com/sul-dlss/digital_stacks_rails/pull/45/files#diff-fe961cf75b59e6eee79d272c43d55a8cR40)).

ndushay commented on app/models/stacks_media_token.rb:
```diff
@@ -0,0 +1,98 @@
+    Settings.stream.max_token_age.to_i.seconds
+  end
+
+  def self.create_from_hash(smt_hash)
```
sadly, I don't know what "smt" in "smt_hash" stands for.   Also, "create_from_hash" doesn't tell me what it's creating.  Maybe just name it "the thing" oh naming-expert-appointed-by-sandi-metz ?

ndushay replied:
I think I prefer "token" over "smt"  as it wasn't immediately obvious to me what "smt" was abbreviating.  Minor quibble, can be ignored.

jmartin-sul replied:
yeah, i think i also like `token_hash` better, will change.

ndushay commented on spec/controllers/media_controller_spec.rb:
```diff
@@ -115,4 +121,42 @@
+      get :verify_token, token_string: encrypted_token, id: id, file_name: file_name, user_ip_addr: '192.168.1.101'
+      expect(response.body).to eq 'invalid token'
+      expect(response.status).to eq 403
+    end
```
should there be a test in this block for rejecting a token because it's too old ?   when the timestamp is invalid?  etc?

jmartin-sul replied:
yeah, actually, the better way to do this would probably be to test that `token_still_valid?` calls `StacksMediaToken.verify_encrypted_token_string` with the expected params, and that `verify_token` does the right thing with the result of `token_still_valid?` (by mocking the `token_still_valid?` call).  the specific validity conditions are tested in `stacks_media_token_spec.rb`, which seems correct/sufficient (since that's where the validity checking logic lives).

ndushay replied:
It could be that all of this could be a single test to show that `verify_token` calls private method `token_still_valid?`  Then and ensure that `token_still_valid?`  calls   the right method in  StacksMediaToken, and that all the tests are done for THAT method.

ndushay commented on app/models/stacks_media_token.rb:
```diff
@@ -0,0 +1,98 @@
+  #  * it hasn't yet expired
+  # NOTE: since this method does an expiry check at the time it's run, its result can
+  # become stale, and should not be cached.
+  def token_still_valid_for?(expected_id, expected_file_name, expected_user_ip_addr)
```
would this be better named, simply, `token_valid?`

ndushay commented on app/controllers/media_controller.rb:
```diff
@@ -47,7 +58,31 @@ def stacks_media_stream_params
     @media ||= StacksMediaStream.new(stacks_media_stream_params)
   end
+
+  def token_still_valid?(token_string, expected_id, expected_file_name, expected_user_ip_addr)
```
would this be better named `token_valid?`

jmartin-sul replied:
sure, i'd do that if your preference for the change was strong.  i like the `still` since it implies that the token won't be valid forever, and was actually debating renaming the `verify_encrypted_token_string` method along those lines, but couldn't think of a better name for that one.

ndushay replied:
I prefer `token_valid?` ;  the fact the the token verification has a time basis is opaque at this point.

ndushay replied:
and now I'm wondering if splitting this method out is better than just calling StacksMediaToken method inside `verify_token` method.   I'm okay with it split out like this, tho.

jkeck commented on app/models/stacks_media_token.rb:
```diff
@@ -0,0 +1,98 @@
+  end
+
+  def self.create_from_hash(smt_hash)
+    new smt_hash[:id], smt_hash[:file_name], smt_hash[:user_ip_addr]
```
Does this provide a benefit over the initializer taking a hash and simply calling `new` in `create_from_encrypted_string`?

jmartin-sul replied:
i like that `new` calls out the expected parameters, since we decided to forgo allowing the token payload to be an arbitrary hash.  but i can't really think of any benefits other than that.

ndushay commented on app/models/stacks_media_token.rb:
```diff
@@ -0,0 +1,98 @@
+  # MessageVerifier#verify, the latter from MessageEncryptor#_decrypt, both called by
+  # MessageEncryptor#decrypt_and_verify).
+  def self.create_from_encrypted_string(encrypted_string)
+    smt_hash = encryptor.decrypt_and_verify encrypted_string
```
I think here, this *could* be `token_hash` or even just `hash` since it is used immediately on the next line.  A quibble, can be ignored.

ndushay commented on spec/models/stacks_media_token_spec.rb:
```diff
@@ -0,0 +1,78 @@
+    it 'returns false if the IP addresses do not match' do
+      expect(subject.token_still_valid_for?(id, file_name, '192.168.1.101')).to eq false
+    end
+
```
test for returns false if it's missing one of the required params?  If there is an invalid value for params?

jmartin-sul replied:
well, i don't think a param can be missing, since that'd be a wrong # of args error (since they're method call params, not keys in a hash).  and this section tests each possible thing that can mismatch, and matching values are the only thing that the check cares about.  

if we wanted to be more stringent in what was allowed, i'd prefer to add that sort of type checking at token creation time.  then you're still just looking for mismatches when you're doing the check (but not, e.g., allowing `blahblahblah` in the IP address field when the token is instantiated).

ndushay replied:
not missing, but an empty string?

ndushay commented:
New batch of comments:

- should any of the StacksMediaToken methods be private?   Is the API for that class truly all the methods?  There is a way to create private class methods that @jkeck showed me.

Also:
- some suggestions for tests to be added
- a couple of method rename suggestions
- minor suggestion about some method internal variable names -- it's okay if you disagree.

ndushay commented on app/controllers/media_controller.rb:
```diff
@@ -19,18 +19,29 @@ def stream
     respond_to do |format|
       format.m3u8 do
-        redirect_to @media.to_playlist_url
+        redirect_to "#{@media.to_playlist_url}?token_string=#{media_token}"
```
@jkeck  for some reason, this feels a little clunky -- we're doing a GET call and including the params in the redirect string.   It's correct, and I can certainly live with it, but wondering if there's a more elegant way to redirect to an external url with GET and param

jkeck replied:
Since `@media.to_playlist_url` is a string URL to an external resource (and not a hash of params to an internal resource) I can't think of a more elegant solution off the top of my head.

More than happy to look at potential alternative solutions.

ndushay replied:
not worth the research project.  Was just curious.

jmartin-sul replied:
yeah, i think jessie and i both felt that it looked a bit inelegant, but not terrible, and seems like everyone can live with it (and no one seems to have an obvious better solution).

jmartin-sul commented:
should hopefully have the new batch of concerns covered:
* added the suggested tests (`StacksMediaToken` now tests that expired tokens are seen as invalid; also added tests to ensure that `MediaController` is calling its methods and the `StacksMediaToken` methods as expected for token generation/verfication).
* added Rails-y field validation for `StacksMediaToken` fields, added tests for said validation.  token creation will fail if input is invalid.
* added comments explaining our assumptions about crypto and implementation when verifying a token.
* incorporated suggested method and variable renaming.

i also took @ndushay's suggestion about making some of the `StacksMediaToken` methods private.  i've left this as a fourth commit since my attempt at that might not actually be what was hoped for.  then it'd be easy to revert that change.  but if that's agreeable, i'm happy to squash it into the first commit to keep the history cleaner.

ndushay commented:
Looks great.  I have but 2 quibbles:

- in MediaStacksToken, perhaps the method should be named `token_valid?`  rather than `token_still_valid_for?`   Same logic applies:  the test is simply if the token is valid or not, for all aspects of the token.

- in MediaStacksToken, an encrypted token will always be a string, no?  So drop the _string suffix on the method name and variable names?

ndushay commented on app/models/stacks_media_token.rb:
```diff
@@ -0,0 +1,120 @@
+  end
+
+  # this is how a string is checked to confirm whether it represents a still-valid token minted by stacks
+  def self.verify_encrypted_token_string(encrypted_string, expected_id, expected_file_name, expected_user_ip_addr)
```
drop the _string suffix?   an encrypted token will always be a string, no?

jmartin-sul commented:
sure.  ok with the way the StacksMediaToken helper methods were made private?  if so, will squash that in first, since that'd make it easier to do the renaming.

ndushay commented:
private class methods in StacksMediaToken look fine to me, but @jkeck might prefer the class << self syntax.  The latter syntax would make their privateness more skimmable, but I don't have strong feelings.

jmartin-sul commented:
i have a slight preference for the current syntax, but waffled on it myself yesterday, and am totally happy to switch to the `class << self` style if that's the consensus.

jmartin-sul commented:
chatted w/ @jkeck a bit just now, sounds like he has no strong pref and is :ok_hand: with usage of `private_class_method` when the list of methods to be delineated as public/private is short (and it's just three short methods in this PR):

from jessie:
> I don’t have a super strong preference, I typically make a judgement call based on the code in the class (e.g. If I’m organizing lots of class methods into public and private, sometimes a clear delineation of them w/ a private is nice, and not having the duplicated private_class_method calls).  When just trying to make a few class methods private (that doesn’t necessarily fit into a good organizational context), I typically go w/ private_class_method.
